monai/losses/unified_focal_loss.py

```diff
@@ -0,0 +1,248 @@
+
+    - "Unified Focal Loss: Generalising Dice and Cross Entropy-based Losses to Handle Class Imbalanced Medical Image Segmentation",
+    Michael Yeung, Computerized Medical Imaging and Graphics
+    """
+
+    def __init__(
+        self,
+        include_background: bool = True,
+        to_onehot_y: bool = False,
+        delta: float = 0.7,
+        gamma: float = 0.75,
+        epsilon: float = 1e-7,
+        reduction: Union[LossReduction, str] = LossReduction.MEAN,
+    ) -> None:
+        """
+        Args:
+            include_background (bool, optional): If False channel index 0 (background category) is excluded from the calculation.
+            Defaults to True.
+            to_onehot_y (bool, optional): whether to convert `y` into the one-hot format. Defaults to False.
+            delta (float, optional): weight of the backgrand. Defaults to 0.7.
+            gamma (float, optional): value of the exponent gamma in the definition of the Focal loss  . Defaults to 0.75.
+            epsilon (float, optional): it's define a very small number each time. simmily smooth value. Defaults to 1e-7.
+        """
+        super().__init__(reduction=LossReduction(reduction).value)
+        self.include_background = include_background
+        self.to_onehot_y = to_onehot_y
+        self.delta = delta
+        self.gamma = gamma
+        self.epsilon = epsilon
+
+    def forward(self, y_pred: torch.Tensor, y_true: torch.Tensor) -> torch.Tensor:
+        n_pred_ch = y_pred.shape[1]
+
+        if self.to_onehot_y:
+            if n_pred_ch == 1:
+                warnings.warn("single channel prediction, `to_onehot_y=True` ignored.")
+            else:
+                y_true = one_hot(y_true, num_classes=n_pred_ch)
+
+        if not self.include_background:
+            if n_pred_ch == 1:
+                warnings.warn("single channel prediction, `include_background=False` ignored.")
+            else:
+                # if skipping background, removing first channel
+                y_true = y_true[:, 1:]
+                y_pred = y_pred[:, 1:]
+
+        if y_true.shape != y_pred.shape:
+            raise ValueError(f"ground truth has different shape ({y_true.shape}) from input ({y_pred.shape})")
+
+        # clip the prediction to avoid NaN
+        y_pred = torch.clamp(y_pred, self.epsilon, 1.0 - self.epsilon)
+        axis = list(range(2, len(y_pred.shape)))
+
+        # Calculate true positives (tp), false negatives (fn) and false positives (fp)
+        tp = torch.sum(y_true * y_pred, dim=axis)
+        fn = torch.sum(y_true * (1 - y_pred), dim=axis)
+        fp = torch.sum((1 - y_true) * y_pred, dim=axis)
+        dice_class = (tp + self.epsilon) / (tp + self.delta * fn + (1 - self.delta) * fp + self.epsilon)
+
+        # Calculate losses separately for each class, enhancing both classes
+        back_dice = 1 - dice_class[:, 0]
+        fore_dice = (1 - dice_class[:, 1]) * torch.pow(1 - dice_class[:, 1], -self.gamma)

```

Comment:
looks like this is for binary segmentation only, so probably `include_background=False` shouldn't be provided as an option and the input must have 2 channels. the docstring should also clarify these.